### PR TITLE
Use custom DB tables

### LINKED ISSUE #98: Use custom DB tables
from here https://wordpress.org/support/topic/przechowywanie-danych-w-wp_postmeta/#post-17998943

- store history not in post_meta but in custom tables
- add migration logic for backward comp.

### PR DESCRIPTION
**Pull request by kkarpieszuk**
# Database Migration: Move Price History from Post Meta to Custom Tables

## Summary

Fixes #98 

This PR implements a comprehensive database migration system to move price history data from WordPress post_meta to dedicated custom database tables, significantly improving query performance and scalability.

### New Features

- **Custom Database Tables**: Created two new tables (`wp_wc_price_history` and `wp_wc_price_history_meta`) for storing price history
- **Migration System**: Automated AJAX-based batch migration from post_meta to tables
- **Progress Tracking**: Real-time progress bar showing migration status
- **Migration Notices**: Admin notices similar to WooCommerce database update flow
- **Backward Compatibility**: Seamless fallback to post_meta for existing installations

#### Database Schema

**Main Table: `wp_wc_price_history`**
- Stores price history with support for both regular and sale prices
- Includes `previous_price` and `previous_sale_price` columns for price comparison
- UNIQUE constraint on (`product_id`, `date_gmt`) prevents duplicate entries
- Optimized indexes for efficient querying

**Meta Table: `wp_wc_price_history_meta`**
- Stores additional metadata about price changes
- Supports metadata like `was_on_sale` and `change_reason` for manual price edits

#### Migration Process

1. Automatic detection of products with existing price history in post_meta
2. AJAX-based batch processing (10 products per batch)
3. Real-time progress bar with percentage and product count
4. Safeguards against duplicate migration (application + database level)
5. Preserves post_meta data as backup

#### Key Files Created

- `app/PriorPrice/Database/Install.php` - Table creation and schema management
- `app/PriorPrice/Database/DbMigration.php` - Migration logic and batch processing
- `app/PriorPrice/Database/AdminNotice.php` - Admin UI for migration notices
- `app/PriorPrice/HistoryStorageTable.php` - New table-based storage implementation
- `assets/js/migration.js` - AJAX migration handlers with progress tracking
- `assets/css/migration.css` - Styling for migration UI

#### Key Files Modified

- `app/PriorPrice/HistoryStorage.php` - Adapter pattern supporting both post_meta and tables
- `app/PriorPrice/Ajax.php` - Added migration AJAX endpoint
- `app/PriorPrice/AdminAssets.php` - Enqueues migration scripts and styles
- `app/PriorPrice/Hooks.php` - Registers migration classes
- `plugin.php` - Activation hooks and table creation

## Benefits

1. **Performance**: Database queries are significantly faster than post_meta lookups
2. **Scalability**: Better handling of large product catalogs with extensive history
3. **Flexibility**: Added columns for previous prices enable better price comparison features
4. **User Experience**: Smooth migration process with clear progress indicators
5. **Data Integrity**: Multiple safeguards prevent duplicate or lost data during migration

### Manual Testing Steps

1. Install/activate plugin on a site with existing products and price history
2. Verify migration notice appears in admin panel
3. Click "Update Database" button
4. Observe progress bar updating in real-time
5. Verify migration completes successfully
6. Check that price history still displays correctly on frontend
7. Verify post_meta data remains intact as backup

### Areas to Test

- Migration of products with extensive price history
- Migration of products with varying price changes
- Handling of products on sale vs regular prices
- Edge cases: products with no price history, zero prices, etc.
- Frontend display of migrated price history

## Database Migration Status

The migration system tracks status using WordPress options:
- `wc_price_history_db_version` - Current database version (2.0.0)
- `wc_price_history_migration_status` - Status (pending/in_progress/completed/not_needed)
- `wc_price_history_migration_total` - Total products to migrate
- `wc_price_history_migration_processed` - Products migrated so far
- `wc_price_history_migrated_products` - Array of migrated product IDs

## Backward Compatibility

- Existing installations continue working with post_meta until migration is triggered
- Plugin automatically detects whether to use tables or post_meta
- Post_meta data is preserved as backup after migration
- Can handle rollback if needed

### Admin Notice States

**Pending**
- Orange/warning colored box
- "WC Price History database update required"
- Call to action button to start migration

**In Progress**
- Blue/info colored box
- Real-time progress bar showing percentage
- Displays "Migrating products: X / Y (Z%)"

**Completed**
- Green/success colored box
- Shows total products migrated
- Dismissible notice

## Future Enhancements

- WP-CLI command for migration
- Ability to manually edit historical prices with reasons
- Export/import functionality for price history
- Advanced price comparison reports
- API endpoints for price history data

## Breaking Changes

None - fully backward compatible.

## Database Version

- Current: 2.0.0
- Previous: 1.x (post_meta based)

## Documentation

- User manual updated with migration information
- Code comments added for all new classes and methods
- README updated with database schema details

## Related Issues

Closes #[issue_number]

## Checklist

- [x] Database tables created with proper schema
- [x] Migration system implemented
- [x] AJAX batch processing working
- [x] Progress bar functionality
- [x] Admin notices displayed
- [x] Backward compatibility maintained
- [x] Code follows WordPress PHP coding standards
- [x] No linter errors
- [x] Files properly namespaced
- [x] Unit tests added/updated
- [x] Documentation updated
- [x] Manual testing completed

---

> [!NOTE]
> Introduces a table-backed storage layer for price history and an in-dashboard migration flow.
> 
> - Adds `wc_price_history` and `wc_price_history_meta` tables with install/version checks and activation hook
> - New migration system: AJAX batch processing, locking, progress tracking, and admin notices; resumes on reload
> - `HistoryStorage` now adapts between legacy post_meta and new `HistoryStorageTable`; optional flag to force legacy
> - Ensures tables exist on load; updates settings page with storage status; enqueues migration JS/CSS
> - CI: adds PHPStan and PHPUnit GitHub Actions; docs and utility scripts for testing/migration
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6b1c7f99286932795de4005e412b84b0d5c12f3f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by kkarpieszuk** [comment]
- [x] na dużej bazie zaobserwowalem ze migracja sie nie konczy. ostatnie kilka produktow (mniej niż batch) się chyba nie migrują. pasek postepu stoi
- [x] ad. wyżej: po odswiezeniu wyswietla sie notice please "Please wait while we update your database..." i nic sie nie dzieje. dodać opcje wymuszenia zakonczenia migracji w takich przypadkach
- [ ] zmien telsty wyswietlane podczas migracji, są zbyt techniczne

File: plugin.php [hunk 1]
```diff
@@ -42,6 +47,13 @@ function get_wc_price_history_version(): string {
 			</div>
 			<?php
 		} );
+		return;
+	}
+
+	// Check if database tables need to be created or updated.
+	if ( ! Install::tables_exist() ) {
+		Install::create_tables();
+		Install::update_db_version();
```
[the remark this answers is not in the document]
Comment (kkarpieszuk) [reply plugin.php]: it must be like this in case someone updated plugin by altering files only. activation hook not started

File: app/PriorPrice/HistoryStorage.php [hunk 2]
```diff
@@ -154,11 +211,17 @@ public function add_price( int $product_id, float $new_price, bool $on_change_on
...
+		if ( $this->should_use_tables() ) {
+			return $this->table_storage->add_first_price( $product_id, $price );
+		}
 
+		// Legacy post_meta implementation.
 		if ( $price <= 0 ) {
 			return 0;
 		}
 
+		$history = [];
```
[the remark this answers is not in the document]
Comment (kkarpieszuk) [reply app/PriorPrice/HistoryStorage.php:236]: this is ok
